File: handlers/pjsk/event.go
package pjsk

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	htmltemplate "html/template"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
)

var validServers = map[string]bool{
	"jp": true, "cn": true, "en": true, "tw": true, "kr": true,
}

type pjskEvent struct {
	ID              int    `json:"id"`
	EventType       string `json:"eventType"`
	Name            string `json:"name"`
	AssetbundleName string `json:"assetbundleName"`
	StartAt         int64  `json:"startAt"`
	AggregateAt     int64  `json:"aggregateAt"`
	ClosedAt        int64  `json:"closedAt"`
	Unit            string `json:"unit"`
}

type pjskEventCard struct {
	CardID  int `json:"cardId"`
	EventID int `json:"eventId"`
}

type pjskCard struct {
	ID              int    `json:"id"`
	CharacterID     int    `json:"characterId"`
	CardRarityType  string `json:"cardRarityType"`
	Attr            string `json:"attr"`
	Prefix          string `json:"prefix"`
	AssetbundleName string `json:"assetbundleName"`
}

// cardDisplay 传给模板的卡面展示数据
type cardDisplay struct {
	Prefix    string
	Rarity    string
	Attr      string
	Thumbnail htmltemplate.URL
	Frame     htmltemplate.URL // 边框 data URL
	AttrIcon  htmltemplate.URL // 属性图标 data URL
	Stars     []int            // 星星 x 坐标列表
	StarIcon  htmltemplate.URL // 星星/birthday 图标 data URL
}

var eventTypeNames = map[string]string{
	"marathon":          "马拉松",
	"cheerful_carnival": "欢乐嘉年华",
	"world_bloom":       "世界绽放",
	"world_link":        "世界连结",
}

var unitNames = map[string]string{
	"idol":           "MORE MORE JUMP!",
	"light_sound":    "Leo/need",
	"school_refusal": "25时、ナイトコードで。",
	"street":         "Vivid BAD SQUAD",
	"theme_park":     "ワンダーランズ×ショウタイム",
	"none":           "混合",
}

var serverNames = map[string]string{
	"jp": "日服", "cn": "国服", "en": "国际服", "tw": "台服", "kr": "韩服",
}

var rarityNames = map[string]string{
	"rarity_1":        "★",
	"rarity_2":        "★★",
	"rarity_3":        "★★★",
	"rarity_4":        "★★★★",
	"rarity_birthday": "Birthday",
}

var attrNames = map[string]string{
	"cool":       "Cool",
	"cute":       "Cute",
	"happy":      "Happy",
	"mysterious": "Mysterious",
	"pure":       "Pure",
}

// --- 卡面 SVG 静态资源 (启动时加载为 data URL) ---

const cardAssetDir = "static/pjsk/card"

var (
	frameDataURLs    map[string]htmltemplate.URL // rarity -> frame data URL
	attrIconDataURLs map[string]htmltemplate.URL // attr -> icon data URL
	starDataURL      htmltemplate.URL
	birthdayDataURL  htmltemplate.URL
)

func loadPNGAsDataURL(path string) htmltemplate.URL {
	data, err := os.ReadFile(path)
	if err != nil {
		return htmltemplate.URL("")
	}
	return htmltemplate.URL("data:image/png;base64," + base64.StdEncoding.EncodeToString(data))
}

func init() {
	frameDataURLs = map[string]htmltemplate.URL{
		"rarity_1":        loadPNGAsDataURL(filepath.Join(cardAssetDir, "cardFrame_S_1.png")),
		"rarity_2":        loadPNGAsDataURL(filepath.Join(cardAssetDir, "cardFrame_S_2.png")),
		"rarity_3":        loadPNGAsDataURL(filepath.Join(cardAssetDir, "cardFrame_S_3.png")),
		"rarity_4":        loadPNGAsDataURL(filepath.Join(cardAssetDir, "cardFrame_S_4.png")),
		"rarity_birthday": loadPNGAsDataURL(filepath.Join(cardAssetDir, "cardFrame_S_bd.png")),
	}
	attrIconDataURLs = map[string]htmltemplate.URL{
		"cool":       loadPNGAsDataURL(filepath.Join(cardAssetDir, "icon_attribute_cool.png")),
		"cute":       loadPNGAsDataURL(filepath.Join(cardAssetDir, "icon_attribute_cute.png")),
		"happy":      loadPNGAsDataURL(filepath.Join(cardAssetDir, "icon_attribute_happy.png")),
		"mysterious": loadPNGAsDataURL(filepath.Join(cardAssetDir, "icon_attribute_mysterious.png")),
		"pure":       loadPNGAsDataURL(filepath.Join(cardAssetDir, "icon_attribute_pure.png")),
	}
	starDataURL = loadPNGAsDataURL(filepath.Join(cardAssetDir, "rarity_star_normal.png"))
	birthdayDataURL = loadPNGAsDataURL(filepath.Join(cardAssetDir, "rarity_birthday.png"))
}

// starPositions 根据稀有度返回星星的 x 坐标列表
func starPositions(rarity string) []int {
	switch rarity {
	case "rarity_1":
		return []int{10}
	case "rarity_2":
		return []int{10, 36}
	case "rarity_3":
		return []int{10, 36, 62}
	case "rarity_4":
		return []int{10, 36, 62, 88}
	case "rarity_birthday":
		return []int{10}
	default:
		return nil
	}
}

// --- 数据查询（基于 masterdata 缓存）---

// findEvent 从缓存的 events.json 中查找活动
func findEvent(server string, eventID int) (*pjskEvent, error) {
	data, err := ReadCachedJSON(server, "events.json")
	if err != nil {
		return nil, err
	}

	var events []pjskEvent
	if err := json.Unmarshal(data, &events); err != nil {
		return nil, fmt.Errorf("解析活动数据失败: %w", err)
	}

	for i := range events {
		if events[i].ID == eventID {
			return &events[i], nil
		}
	}
	return nil, fmt.Errorf("未找到活动 ID: %d (服务器: %s)", eventID, server)
}

func latestEventByTime(server string) (*pjskEvent, error) {
	data, err := ReadCachedJSON(server, "events.json")
	if err != nil {
		return nil, err
	}

	var events []pjskEvent
	if err := json.Unmarshal(data, &events); err != nil {
		return nil, fmt.Errorf("解析活动数据失败: %w", err)
	}
	if len(events) == 0 {
		return nil, fmt.Errorf("活动数据为空 (服务器: %s)", server)
	}

	latest := &events[0]
	for i := 1; i < len(events); i++ {
		if isNewerEvent(events[i], *latest) {
			latest = &events[i]
		}
	}

	if latest.ID <= 0 {
		return nil, fmt.Errorf("未找到有效活动 ID (服务器: %s)", server)
	}
	return latest, nil
}

func isNewerEvent(a, b pjskEvent) bool {
	aTime := latestEventTime(a)
	bTime := latestEventTime(b)
	if aTime != bTime {
		return aTime > bTime
	}
	if a.StartAt != b.StartAt {
		return a.StartAt > b.StartAt
	}
	if a.AggregateAt != b.AggregateAt {
		return a.AggregateAt > b.AggregateAt
	}
	if a.ClosedAt != b.ClosedAt {
		return a.ClosedAt > b.ClosedAt
	}
	return a.ID > b.ID
}

func latestEventTime(e pjskEvent) int64 {
	if e.StartAt > 0 {
		return e.StartAt
	}
	if e.AggregateAt > 0 {
		return e.AggregateAt
	}
	return e.ClosedAt
}

// findEventCards 查找活动关联的卡面信息
func findEventCards(server string, eventID int) []cardDisplay {
	ecData, err := ReadCachedJSON(server, "eventCards.json")
	if err != nil {
		return nil
	}
	var eventCards []pjskEventCard
	if err := json.Unmarshal(ecData, &eventCards); err != nil {
		return nil
	}

	// 收集该活动的 cardID
	var cardIDs []int
	for _, ec := range eventCards {
		if ec.EventID == eventID {
			cardIDs = append(cardIDs, ec.CardID)
		}
	}
	if len(cardIDs) == 0 {
		return nil
	}

	cData, err := ReadCachedJSON(server, "cards.json")
	if err != nil {
		return nil
	}
	var cards []pjskCard
	if err := json.Unmarshal(cData, &cards); err != nil {
		return nil
	}

	// 建 cardID -> card 索引
	cardMap := make(map[int]*pjskCard, len(cards))
	for i := range cards {
		cardMap[cards[i].ID] = &cards[i]
	}

	var result []cardDisplay
	for _, cid := range cardIDs {
		card, ok := cardMap[cid]
		if !ok {
			continue
		}
		thumb := downloadAssetByLabel(server, "card:thumbnail:"+card.AssetbundleName+":normal")

		rarity := card.CardRarityType
		if name, ok := rarityNames[rarity]; ok {
			rarity = name
		}
		attr := card.Attr
		if name, ok := attrNames[attr]; ok {
			attr = name
		}

		// SVG 资源
		frame := frameDataURLs[card.CardRarityType]
		attrIcon := attrIconDataURLs[card.Attr]
		stars := starPositions(card.CardRarityType)
		starIcon := starDataURL
		if card.CardRarityType == "rarity_birthday" {
			starIcon = birthdayDataURL
		}

		result = append(result, cardDisplay{
			Prefix:    card.Prefix,
			Rarity:    rarity,
			Attr:      attr,
			Thumbnail: thumb,
			Frame:     frame,
			AttrIcon:  attrIcon,
			Stars:     stars,
			StarIcon:  starIcon,
		})
	}
	return result
}

func formatMillisTime(ms int64) string {
	if ms <= 0 {
		return ""
	}
	return time.Unix(ms/1000, 0).Local().Format("2006-01-02 15:04:05")
}

func eventTypeName(t string) string {
	if name, ok := eventTypeNames[t]; ok {
		return name + " (" + t + ")"
	}
	return t
}

func eventStatus(startAt, closedAt int64) string {
	now := time.Now().UnixMilli()
	if now < startAt {
		return "未开始"
	}
	if now > closedAt {
		return "已结束"
	}
	return "进行中"
}

func unitName(u string) string {
	if name, ok := unitNames[u]; ok {
		return name
	}
	return u
}

// eventProgress 返回活动进度百分比（0~100），未开始返回 0，已结束返回 100
func eventProgress(startAt, closedAt int64) float64 {
	now := time.Now().UnixMilli()
	if now <= startAt {
		return 0
	}
	if now >= closedAt {
		return 100
	}
	total := float64(closedAt - startAt)
	elapsed := float64(now - startAt)
	return elapsed / total * 100
}

func renderEventError(c *gin.Context, errMsg string) {
	c.HTML(http.StatusOK, "pjsk/event", gin.H{
		"Error": errMsg,
	})
}

func EventHandler(c *gin.Context) {
	server := c.DefaultQuery("server", "jp")
	if !validServers[server] {
		renderEventError(c, "无效的服务器参数，支持: jp, cn, en, tw, kr")
		return
	}

	idStr := c.Query("id")
	if idStr == "" {
		c.Redirect(http.StatusFound, "/pjsk/event/current?server="+server)
		return
	}

	eventID, err := strconv.Atoi(idStr)
	if err != nil {
		renderEventError(c, "无效的活动 ID: "+idStr)
		return
	}

	target, err := findEvent(server, eventID)
	if err != nil {
		renderEventError(c, err.Error())
		return
	}

	// 下载背景图、徽标、横幅
	bgDataURL := downloadAssetByLabel(server, "event:background:"+target.AssetbundleName)
	logoDataURL := downloadAssetByLabel(server, "event:logo:"+target.AssetbundleName)
	bannerDataURL := downloadAssetByLabel(server, "event:banner:"+target.AssetbundleName)

	// 查找活动关联卡面
	cards := findEventCards(server, eventID)

	c.HTML(http.StatusOK, "pjsk/event", gin.H{
		"Name":        target.Name,
		"ID":          target.ID,
		"EventType":   eventTypeName(target.EventType),
		"Server":      serverNames[server],
		"ServerKey":   server,
		"Unit":        unitName(target.Unit),
		"Background":  bgDataURL,
		"Logo":        logoDataURL,
		"Banner":      bannerDataURL,
		"StartAt":     formatMillisTime(target.StartAt),
		"ClosedAt":    formatMillisTime(target.ClosedAt),
		"AggregateAt": formatMillisTime(target.AggregateAt),
		"Status":      eventStatus(target.StartAt, target.ClosedAt),
		"Progress":    fmt.Sprintf("%.1f", eventProgress(target.StartAt, target.ClosedAt)),
		"Cards":       cards,
	})
}

func CurrentEventHandler(c *gin.Context) {
	server := c.DefaultQuery("server", "jp")
	if !validServers[server] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "无效的服务器参数，支持: jp, cn, en, tw, kr"})
		return
	}

	latest, err := latestEventByTime(server)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "获取最新活动失败: " + err.Error()})
		return
	}

	redirectURL := fmt.Sprintf("/pjsk/event?id=%d&server=%s", latest.ID, server)
	c.Redirect(http.StatusFound, redirectURL)
}

File: src/handlers/pjsk/event.rs
use crate::handlers::pjsk::asset_source::download_asset_by_label;
use crate::handlers::pjsk::assets::read_cached_json;
use crate::handlers::pjsk::card::star_positions;
use crate::handlers::pjsk::{SERVER_NAMES, VALID_SERVERS};
use crate::handlers::render_html;
use axum::{extract::Query, http::StatusCode, response::IntoResponse};
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Debug)]
pub struct EventQuery {
    pub id: Option<String>,
    pub server: Option<String>,
}

#[derive(Serialize, Clone)]
#[allow(non_snake_case)]
pub struct CardView {
    pub ID: i32,
    pub Prefix: String,
    pub Thumbnail: String,
    pub Frame: String,
    pub AttrIcon: String,
    pub Stars: Vec<i32>,
    pub StarIcon: String,
}

#[derive(Serialize, Clone)]
#[allow(non_snake_case)]
pub struct EventPageData {
    pub ID: i32,
    pub Name: String,
    pub EventType: String,
    pub StartAt: String,
    pub AggregateAt: String,
    pub ClosedAt: String,
    pub Server: String,
    pub ServerKey: String,
    pub Banner: String,
    pub Logo: String,
    pub Status: &'static str,
    pub Progress: String,
    pub Cards: Vec<CardView>,
    pub FooterExtra: String,
}

#[derive(Serialize)]
#[allow(non_snake_case)]
pub struct EventResponse {
    pub Event: Option<EventPageData>,
    pub Error: Option<String>,
}

#[derive(Deserialize, Debug)]
#[allow(non_snake_case)]
struct EventEntry {
    id: i32,
    name: String,
    eventType: String,
    assetbundleName: String,
    startAt: i64,
    aggregateAt: i64,
    closedAt: i64,
}

#[derive(Deserialize, Debug)]
#[allow(non_snake_case)]
struct CardEntry {
    id: i32,
    prefix: String,
    assetbundleName: String,
    cardRarityType: String,
    attribute: String,
}

#[derive(Deserialize, Debug)]
#[allow(non_snake_case)]
struct EventCardsEntry {
    eventId: i32,
    cardId: i32,
}

fn format_millis_time(ms: i64) -> String {
    crate::pkg::timefmt::format_unix_millis(ms)
}

fn event_status(start: i64, close: i64) -> &'static str {
    let now = chrono::Local::now().timestamp_millis();
    if now < start {
        "未开始"
    } else if now > close {
        "已结束"
    } else {
        "进行中"
    }
}

fn event_progress(start: i64, close: i64) -> f64 {
    let now = chrono::Local::now().timestamp_millis();
    if now < start {
        return 0.0;
    }
    if now > close {
        return 100.0;
    }
    let total = (close - start) as f64;
    if total <= 0.0 {
        return 100.0;
    }
    ((now - start) as f64 / total * 100.0).min(100.0).max(0.0)
}

pub async fn event_handler(Query(q): Query<EventQuery>) -> impl IntoResponse {
    let server = q.server.unwrap_or_else(|| "jp".to_string());
    if !VALID_SERVERS.contains(&server) {
        return render_html(
            "pjsk/event.html",
            EventResponse {
                Event: None,
                Error: Some("无效的服务器参数，支持: jp, cn, en, tw, kr".to_string()),
            },
        )
        .into_response();
    }

    let event_id_str = q.id.unwrap_or_default().trim().to_string();
    if event_id_str.is_empty() {
        return render_html(
            "pjsk/event.html",
            EventResponse {
                Event: None,
                Error: Some("缺少活动 ID 参数".to_string()),
            },
        )
        .into_response();
    }

    let event_id: i32 = match event_id_str.parse() {
        Ok(id) if id > 0 => id,
        _ => {
            return render_html(
                "pjsk/event.html",
                EventResponse {
                    Event: None,
                    Error: Some("无效的活动 ID".to_string()),
                },
            )
            .into_response();
        }
    };

    // Load event masterdata
    let events_data = match read_cached_json(&server, "events.json") {
        Ok(d) => d,
        Err(e) => {
            return render_html(
                "pjsk/event.html",
                EventResponse {
                    Event: None,
                    Error: Some(e),
                },
            )
            .into_response()
        }
    };
    let events: Vec<EventEntry> = match serde_json::from_slice(&events_data) {
        Ok(evs) => evs,
        Err(e) => {
            return render_html(
                "pjsk/event.html",
                EventResponse {
                    Event: None,
                    Error: Some(format!("解析 events.json 失败: {}", e)),
                },
            )
            .into_response()
        }
    };

    let target = match events.iter().find(|e| e.id == event_id) {
        Some(e) => e,
        None => {
            return render_html(
                "pjsk/event.html",
                EventResponse {
                    Event: None,
                    Error: Some(format!("未找到活动 #{}", event_id)),
                },
            )
            .into_response()
        }
    };

    // Parse event type (marathon, cheerful_carnival etc.)
    let ev_type = match target.eventType.as_str() {
        "marathon" => "马拉松活动",
        "cheerful_carnival" => "欢乐狂欢节活动",
        other => other,
    };

    // Retrieve related cards from eventDeckCards.json & cards.json
    let mut cards = Vec::new();
    if let Ok(deck_data) = read_cached_json(&server, "eventDeckCards.json") {
        if let Ok(deck_cards) = serde_json::from_slice::<Vec<EventCardsEntry>>(&deck_data) {
            let card_ids: Vec<i32> = deck_cards
                .iter()
                .filter(|dc| dc.eventId == event_id)
                .map(|dc| dc.cardId)
                .collect();

            if let Ok(cards_data) = read_cached_json(&server, "cards.json") {
                if let Ok(all_cards) = serde_json::from_slice::<Vec<CardEntry>>(&cards_data) {
                    for cid in card_ids {
                        if let Some(c) = all_cards.iter().find(|c| c.id == cid) {
                            let label = format!("card:thumbnail:{}:normal", c.assetbundleName);
                            let thumb = download_asset_by_label(&server, &label).await;

                            let rarity = &c.cardRarityType;
                            let star_icon = if rarity == "rarity_birthday" {
                                "/static/pjsk/card/rarity_birthday.png".to_string()
                            } else {
                                "/static/pjsk/card/rarity_star_normal.png".to_string()
                            };

                            let frame = format!(
                                "/static/pjsk/card/cardFrame_S_{}.png",
                                match rarity.as_str() {
                                    "rarity_1" => "1",
                                    "rarity_2" => "2",
                                    "rarity_3" => "3",
                                    "rarity_4" => "4",
                                    "rarity_birthday" => "bd",
                                    _ => "1",
                                }
                            );

                            cards.push(CardView {
                                ID: c.id,
                                Prefix: c.prefix.clone(),
                                Thumbnail: thumb,
                                Frame: frame,
                                AttrIcon: format!(
                                    "/static/pjsk/card/icon_attribute_{}.png",
                                    c.attribute
                                ),
                                Stars: star_positions(rarity),
                                StarIcon: star_icon,
                            });
                        }
                    }
                }
            }
        }
    }

    let banner_label = format!("event:banner:{}", target.assetbundleName);
    let banner = download_asset_by_label(&server, &banner_label).await;

    let logo_label = format!("event:logo:{}", target.assetbundleName);
    let logo = download_asset_by_label(&server, &logo_label).await;

    let status = event_status(target.startAt, target.closedAt);
    let progress = format!("{:.1}", event_progress(target.startAt, target.closedAt));

    let page = EventPageData {
        ID: target.id,
        Name: target.name.clone(),
        EventType: ev_type.to_string(),
        StartAt: format_millis_time(target.startAt),
        AggregateAt: format_millis_time(target.aggregateAt),
        ClosedAt: format_millis_time(target.closedAt),
        Server: SERVER_NAMES
            .get(&server)
            .cloned()
            .unwrap_or_else(|| server.to_uppercase()),
        ServerKey: server,
        Banner: banner,
        Logo: logo,
        Status: status,
        Progress: progress,
        Cards: cards,
        FooterExtra: "Powered by Moesekai, Haruki, LunaBot, Uni, & Sekai World<br />".to_string(),
    };

    render_html(
        "pjsk/event.html",
        EventResponse {
            Event: Some(page),
            Error: None,
        },
    )
    .into_response()
}

pub async fn current_event_handler(Query(q): Query<EventQuery>) -> impl IntoResponse {
    let server = q.server.unwrap_or_else(|| "jp".to_string());
    if !VALID_SERVERS.contains(&server) {
        return (StatusCode::BAD_REQUEST, "无效服务器").into_response();
    }

    let events_data = match read_cached_json(&server, "events.json") {
        Ok(d) => d,
        Err(e) => return (StatusCode::INTERNAL_SERVER_ERROR, e).into_response(),
    };
    let events: Vec<EventEntry> = match serde_json::from_slice(&events_data) {
        Ok(evs) => evs,
        Err(e) => {
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("解析 events.json 失败: {}", e),
            )
                .into_response()
        }
    };

    let now = chrono::Local::now().timestamp_millis();
    let current = events
        .iter()
        .find(|e| now >= e.startAt && now <= e.closedAt);

    let redirect_id = match current {
        Some(e) => e.id,
        None => {
            // Find the latest ended event
            events
                .iter()
                .filter(|e| now > e.closedAt)
                .max_by_key(|e| e.closedAt)
                .map(|e| e.id)
                .unwrap_or(1)
        }
    };

    let redirect_url = format!("/pjsk/event?id={}&server={}", redirect_id, server);
    (
        StatusCode::FOUND,
        [(axum::http::header::LOCATION, redirect_url.as_str())],
    )
        .into_response()
}
